server/src/main/java/org/opensearch/index/ShardIndexingPressureTracker.java

```diff
@@ -0,0 +1,223 @@
+
+/**
+ * This class is responsible for all the tracking that needs to be performed at every Shard Level for Indexing Operations on the node.
+ * Info is maintained at the granularity of three kinds of write operation (tasks) on the node i.e. coordinating, primary and replica.
+ * This is useful in evaluating the shard indexing back-pressure on the node, to throttle requests and also to publish runtime stats.
+ *
+ * There can be four kinds of operation tracking on a node which needs to performed for a shard:
+ * 1. Coordinating Operation : To track all the individual shard bulk request on the coordinator node.
+ * 2. Primary Operation : To track all the individual shard bulk request on the primary node.
+ * 3. Replica Operation : To track all the individual shard bulk request on the replica node.
+ * 4. Common Operation : To track values applicable across the specific shard role.
+ *
+ * ShardIndexingPressureTracker therefore provides the construct to track all the write requests targeted for a ShardId on the node,
+ * across all possible transport-write-actions i.e. Coordinator, Primary and Replica.
+ * Tracker is uniquely identified against a Shard-Id on the node. Currently the knowledge of shard roles (such as primary vs replica)
+ * is not explicit to the tracker, and it is able to track different values simultaneously based on the interaction hooks of the
+ * operation type i.e. write-action layers.
+ *
+ * There is room for introducing more unique identity to the trackers based on Shard-Role or Shard-Allocation-Id, but that will also
+ * increase the complexity of handling shard-lister events and handling other race scenarios such as request-draining etc.
+ * To prefer simplicity we have modelled by keeping explicit fields for different operation tracking, while tracker by itself is
+ * agnostic of the actual shard role.
+ */
+public class ShardIndexingPressureTracker {
+
+    private final ShardId shardId;
+    private final AtomicLong primaryAndCoordinatingLimits;
+    private final AtomicLong replicaLimits;
+
+    private final OperationTracker coordinatingOperationTracker = new OperationTracker();
+    private final OperationTracker primaryOperationTracker = new OperationTracker();
+    private final OperationTracker replicaOperationTracker = new OperationTracker();
+    private final CommonOperationTracker commonOperationTracker = new CommonOperationTracker();
+
+    public ShardIndexingPressureTracker(ShardId shardId, long primaryAndCoordinatingLimits, long replicaLimits) {
+        this.shardId = shardId;
+        this.primaryAndCoordinatingLimits = new AtomicLong(primaryAndCoordinatingLimits);
+        this.replicaLimits = new AtomicLong(replicaLimits);
+    }
+
+    public ShardId getShardId() {
+        return shardId;
+    }
+
+    public AtomicLong getPrimaryAndCoordinatingLimits() {
+        return primaryAndCoordinatingLimits;
+    }
+
+    public AtomicLong getReplicaLimits() {
+        return replicaLimits;
+    }
```

Comment:
Don't think this is a good idea to return AtomicLong. Instead let trackers manage the increment/decrement operations and only allow the corresponding long values to be returned

Reply:
The primary reason is the ``AtomicLong`` complex object in itself and exposing all its functionalities would make the class very verbose. For example the caller *(MemoryManager going forward)*  intends to use ``compareAndSet``, ``incrementAndGet``, ``getAndAdd`` ``addAndGet`` and others based on its logic. 
Hence keeping the ``ShardIndexingPressureTracker`` as a simple POJO here which just provides the storage Structure, while oprations/manipulations are handled by ``MemoryManager``

Reply:
The class suffers from lack of concrete ownership, any other class is free to modify tracker as needed which might lead to corruption/bugs. Maintainability might get harder over time.

Reply:
Maybe I was not clear with the above explanation on the responsibilities. The ownership of Tracker will remain with the single class lets say ``MemoryManager`` which will act as an data accessor/modifier for the tracker values, for everybody else in the system. Think of Tracker like a ``DataObject``, while MemoryManager as the ``DataAccessor`` layer for fetch/update operations across the board. 
Now the guarantees around object access to avoid corruption/bugs will be maintained by the MemoryManager. 

To make sure we maintain these guarantees as we build above layers, and not miss on them, I have captured an item in the Meta https://github.com/opensearch-project/OpenSearch/issues/478#issuecomment-813457820 We can bring in more abstractions if needed in future.

Reply:
Based on the conversation, I have ensured that the current interfaces only return immutable types. This will help provide a cleaner perspective and unblock this change. 
We will evolve this in future, further adding more interfaces, as the above layers come up.